File: elixir/lib/2023/04_scratchcards/aoc202304.ex
defmodule AOC2023.Day04 do
  @moduledoc """
  Advent of Code 2023, day 4: Scratchcards.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n")
    |> Enum.map(fn line ->
      line
      |> String.split(":")
      |> Enum.at(1)
      |> String.split("|")
      |> Enum.map(fn tickets -> tickets |> String.split() |> MapSet.new() end)
      |> then(fn [win, own] -> MapSet.intersection(win, own) end)
      |> MapSet.size()
    end)
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    data
    |> Enum.filter(&(&1 > 0))
    |> Enum.map(&(2 ** (&1 - 1)))
    |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    data
    |> Enum.with_index(1)
    |> Enum.filter(fn {win, _} -> win > 0 end)
    |> Enum.reduce(
      1..length(data) |> Enum.map(fn idx -> {idx, 1} end) |> Enum.into(%{}),
      fn {win, id}, counts ->
        (id + 1)..(id + win)
        |> Enum.reduce(counts, &win_new_scratchcards(&1, &2, Map.get(counts, id)))
      end
    )
    |> Map.values()
    |> Enum.sum()
  end

  @doc """
  Update the list of scratchcards with new copies.

  ## Example:

      iex> win_new_scratchcards(3, %{1 => 1, 2 => 2, 3 => 4, 4 => 7, 5 => 1}, 5)
      %{1 => 1, 2 => 2, 3 => 9, 4 => 7, 5 => 1}
  """
  def win_new_scratchcards(id, counts, copies) do
    Map.update(counts, id, nil, fn count -> count + copies end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2023/04_scratchcards/aoc202304.py
"""AoC 4, 2023: Scratchcards."""

# Standard library imports
import pathlib
import sys

# Third party imports
import parse

TICKET_PATTERN = parse.compile("Card {id:3d}: {winning} | {tickets}")


def parse_data(puzzle_input):
    """Parse input."""
    return [
        len(set(match["winning"].split()) & set(match["tickets"].split()))
        for line in puzzle_input.split("\n")
        if (match := TICKET_PATTERN.parse(line))
    ]


def part1(num_winning):
    """Solve part 1."""
    return sum(2 ** (wins - 1) for wins in num_winning if wins > 0)


def part2(num_winning):
    """Solve part 2."""
    counts = [1 for _ in num_winning]
    for index, num in enumerate(num_winning):
        for idx in range(index + 1, index + num + 1):
            counts[idx] += counts[index]
    return sum(counts)


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
